src/daft-dsl/src/functions/scalar.rs

```diff
@@ -40,6 +40,41 @@ impl From<ScalarFunction> for ExprRef {
     }
 }
 
+/// This is a factory for scalar function implementations.
+///
+/// TODO:
+///   Rename to ScalarFunction (or similar) once ScalarFunction is migrated
+///   to ScalarUDF, and update ScalarUDF to ScalarFunctionImpl (or similar).
+///   Then update Expr::Function(ScalarUDF) to Expr::Function(ScalarFunctionFactory)
+///   which will enable *name* resolution within the DSL, but then we now have
+///   the ability for *type* resolution during planning via get_function. We can
+///   build rule-based type resolution at a later time.
+///
+pub trait ScalarFunctionFactory: Send + Sync {
+    /// The name of this function.
+    fn name(&self) -> &'static str;
+
+    /// Any additional aliases for this function.
+    fn aliases(&self) -> &'static [&'static str] {
+        &[]
+    }
+
+    /// Returns a ScalarUDF for the given fields.
+    ///
+    /// Note:
+    ///   When the time comes, we should replace FunctionArgs<ExprRef> with bound ExprRef.
+    ///   This way we have the pair (Expr, Field) so we don't have to keep re-computing
+    ///   expression types each time we resolve a function. At present, I wanted to keep
+    ///   this signature the exact same as function_args_to_field.
+    ///
+    fn get_function(
+        &self,
+        args: FunctionArgs<ExprRef>,
+        schema: &Schema,
+    ) -> DaftResult<Arc<dyn ScalarUDF>>;
```

Comment:
What's an example of what an implementation of this would look like for a non-dynamic function?

Reply:
As Cory mentioned, we can define function args with the `FunctionArgs` derive macro, which could maybe be used instead of this. Just try parsing to each set of args until one succeeds. We can also add type constraints to the macro as well.

The only potential issue with that is if multiple sets of args can be parsed from some input. In that case, defining an explicit `get_function` logic would make it more explicit which implementation should be used.

Reply:
Ah I see. This makes sense. I had no other comments, feel free to merge this.

Note: for the binary function examples you gave, I believe that could be solved with the proper [implementation of `FromLiteral`](https://github.com/Eventual-Inc/Daft/pull/4398/files#diff-be24aaeb60f8fd215f6ae358751cb75c683cba62eae22bbc0053fdb427d14e3cR35-R46) instead of function overloads